### ISSUE TITLE
Mixing concatenating assignment operator and assigment operation results in false positive.

### ISSUE THREAD

**Issue by alexpott** [sirbrillig/phpcs-variable-analysis#183]
Given the following code
```php
  $suggestion = 'block';
  while ($part = array_shift($parts)) {
    $suggestions[] = $suggestion .= '__' . strtr($part, '-', '_');
  }
```
$suggestion will be marked as unused.

**Comment by alexpott** [comment]
Similarly $entry is reported as unused.
```php
    foreach ($this->derivatives as &$entry) {
      $entry += $base_plugin_definition;
    }

    return $this->derivatives;
```

**Comment by jrfnl** [comment]
@alexpott Re the second example: IMO this is correct.
You are only assigning something to `$entry`, but never _using_ the resulting value.

**Comment by kiamlaluno** [comment]
That type of code is enough common, with latest PHP versions. It would be nice if the code would not be reported.

**Comment by alexpott** [comment]
@jrfnl thanks for looking at this issue.

Re the second example - it's a tricky one because I think `$entry` is used. It's a reference to an element in `$this->derivatives` - so `$this->derivatives` is being changed by the `foreach` loop and then `$this->derivatives` is being returned.

**Comment by sirbrillig** [comment]
Thanks for the report! This might be worth reporting as two separate issues unless you see a connection that I'm missing (which is very possible).

I haven't dug into it yet but I'm guessing that in the first case the assignment is masking the assignment to the single scalar. 

In the second case, there's probably a bug with how assignment by reference is being detected? Honestly it's hard for me to grok what's happening there. That code loops through an array of strings and appends another string to each one, modifying the original array items?

**Comment by sirbrillig** [comment]
Another interesting bug I found when researching this:

```php
function concatAndAssignWithUndefinedShift() {
  $suggestions = [];
  while ($part = array_shift($parts)) { // undefined variable parts
    $suggestions[] = $part;
  }
	return $suggestions;
}
```

`$parts` is not marked as undefined.

**Comment by sirbrillig** [comment]
Ok, I know why this is happening. In this example, `$suggestion` is marked as unused.

```php
function assignFunction($parts) {
  $suggestion = 'block';
  $suggestion .= '__';
  return $parts;
}
```

The reason is that even though `$suggestion` is modified, it is never _read_. Changing the example to return the variable removes the warning.

```php
function assignFunction() {
  $suggestion = 'block';
  $suggestion .= '__';
  return $suggestion;
}
```

In the case of your first example, the expression `$suggestions[] = $suggestion .= '__'` is not being counted as a _read_ of `$suggestion`. This is because `$suggestion .= '__'` is only a write but when combined with the surrounding expression it is both a write _and_ a read. The sniff will need to be modified to detect such situations.

**Comment by sirbrillig** [comment]
I think a way to resolve this will be: when marking a variable as a write, also look to see if it's on the right-hand-side of an assignment. If so, also mark that variable as a read.

The difficult thing will be how to determine if we're on the right-hand-side of an assignment. For the above example, it is immediately preceded by an equals sign, but it could also be `$suggestions[] = ( $suggestion .= '__' );` or `$suggestions[] = getPrefix() . $suggestion .= '__';` or probably some more complex expressions.
